mathlibtools/leanproject.py

```diff
@@ -0,0 +1,134 @@
+import sys
+from pathlib import Path
+
+from git.exc import GitCommandError # type: ignore
+import click
+
+from mathlibtools.lib import (LeanProject, log, LeanDirtyRepo,
+    InvalidLeanProject, LeanDownloadError)
+
+# Click aliases from Stephen Rauch at
+# https://stackoverflow.com/questions/46641928
+class CustomMultiCommand(click.Group):
+    def command(self, *args, **kwargs):
+        """Behaves the same as `click.Group.command()` except if passed
+        a list of names, all after the first will be aliases for the first.
+        """
+        def decorator(f):
+            if args and isinstance(args[0], list):
+                _args = [args[0][0]] + list(args[1:])
+                for alias in args[0][1:]:
+                    cmd = super(CustomMultiCommand, self).command(
+                        alias, *args[1:], **kwargs)(f)
+                    cmd.short_help = "Alias for '{}'".format(_args[0])
+            else:
+                _args = args
+            cmd = super(CustomMultiCommand, self).command(
+                *_args, **kwargs)(f)
+            return cmd
+
+        return decorator
+
+def proj():
+    return LeanProject.from_path(Path('.'))
+
+@click.group(cls=CustomMultiCommand)
+def cli():
+    pass
+
+@cli.command()
+@click.argument('path', default='.')
+def new(path: str = '.'):
+    """Create a new Lean project and prepare mathlib.
+
+    If no directory name is given, the current directory is use.
+    """
+    LeanProject.new(Path(path))
+
+@cli.command()
+def add_mathlib():
+    """Add mathlib to the current project."""
+    proj().add_mathlib()
+
+@cli.command(['upgrade-mathlib', 'update-mathlib', 'up'])
+def upgrade_mathlib():
+    """Upgrade mathlib."""
+    try:
+        proj().upgrade_mathlib()
+    except LeanDownloadError:
+        log.error('Failed to fetch mathlib oleans')
+        sys.exit(-1)
+    except Exception as err:
+        log.error(err)
+        sys.exit(-1)
+
+@cli.command()
+def build():
+    """Build the current project."""
+    proj().build()
+
+@cli.command(name='get')
+@click.argument('url')
+@click.argument('target', default='')
```

Comment:
Perhaps we should rename this argument, perhaps simply to `dir`. When I first read this I was confused by the similarity between `target` and `_target`, and wondered if this command was actually for adding dependencies, rather than cloning.

Reply:
I don't mind.